packages/typespec-client-generator-core/src/internal-utils.ts

```diff
@@ -595,3 +607,80 @@ export function findRootSourceProperty(property: ModelProperty): ModelProperty {
   }
   return property;
 }
+
+function getVersioningMutator(
+  context: TCGCContext,
+  service: Namespace,
+  apiVersion: string,
+): unsafe_MutatorWithNamespace {
+  const versionMutator = getVersioningMutators(context.program, service);
+  compilerAssert(
+    versionMutator !== undefined && versionMutator.kind !== "transient",
+    "Versioning service should not get undefined or transient versioning mutator",
+  );
+
+  const mutators = versionMutator.snapshots
+    .filter((snapshot) => apiVersion === snapshot.version.value)
+    .map((x) => x.mutator);
+  compilerAssert(mutators.length === 1, "One api version should not get multiple mutators");
+
+  return mutators[0];
+}
+
+function handleVersioningMutationForGlobalNamespace(context: TCGCContext): Namespace {
+  const globalNamespace = context.program.getGlobalNamespaceType();
+  const allApiVersions = getVersions(context.program, globalNamespace)[1]
+    ?.getVersions()
+    .map((x) => x.value);
+  if (!allApiVersions) return globalNamespace;
+
+  const apiVersion = getValidApiVersion(context, allApiVersions);
+  if (apiVersion === undefined) return globalNamespace;
+
+  const mutator = getVersioningMutator(context, globalNamespace, apiVersion);
+  const subgraph = unsafe_mutateSubgraphWithNamespace(context.program, [mutator], globalNamespace);
+  compilerAssert(subgraph.type.kind === "Namespace", "Should not have mutated to another type");
+  return subgraph.type;
+}
+
+function listAllUserDefinedNamespaces(
+  context: TCGCContext,
+  namespace: Namespace,
+  retval?: Namespace[],
+): Namespace[] {
+  if (!retval) {
+    retval = [];
+  }
+  if (retval.includes(namespace)) return retval;
+  if (getLocationContext(context.program, namespace).type === "project") {
+    retval.push(namespace);
+  }
+  for (const ns of namespace.namespaces.values()) {
+    listAllUserDefinedNamespaces(context, ns, retval);
+  }
+  return retval;
+}
+
+/**
+ * Currently, listServices can only be called from a program instance. This doesn't work well if we're doing mutation,
+ * because we want to just mutate the global namespace once, then find all of the services in the program, since we aren't
+ * able to explicitly tell listServices to iterate over our specific mutated global namespace. We're going to use this function
+ * instead to list all of the services in the global namespace.
+ *
+ * See https://github.com/microsoft/typespec/issues/6247
+ *
+ * @param context
+ */
+export function listAllServiceNamespaces(context: TCGCContext): Namespace[] {
+  const serviceNamespaces: Namespace[] = [];
+  let globalNamespace = context.globalNamespace;
+  if (!globalNamespace) {
+    globalNamespace = handleVersioningMutationForGlobalNamespace(context);
```

Comment:
Should we store the mutated global namespace to `context` or it happens somewhere else? just reading the doc of `TCGCContext.globalNamespace`

```ts
globalNamespace?: Namespace; // the root of all tsp namespaces for this instance. Starting point for traversal, so we don't call mutation multiple times
```

Reply:
in case TCGC is supposed to store it in the context, I wonder why we not store the resolved `serviceNamespaces` array in `context` instead of `globalNamespace`? Computing `serviceNamespaces ` seems to involve some work, but I'm unsure if it always computes the same result.

Reply:
yep `.globalNamespace` is the mutated global namespace. maybe I can change the name